src/generators/utils/pagingOperations.ts

```diff
@@ -0,0 +1,378 @@
+  sourceFile: SourceFile
+) {
+  if (
+    !options.disablePagingAsyncIterators &&
+    operationGroups.some(og => hasAsyncIteratorOperations(og))
+  ) {
+    sourceFile.addImportDeclarations([
+      {
+        namedImports: ["PagedAsyncIterableIterator"],
+        moduleSpecifier: "@azure/core-paging"
+      }
+    ]);
+  }
+}
+
+/**
+ * Checks whether or not an operation group contains any pageable operations
+ * that would need AsyncIterators
+ */
+export function hasAsyncIteratorOperations(
+  operationGroupDetails: OperationGroupDetails
+) {
+  return operationGroupDetails.operations.some(o => o.pagination);
+}
+
+/**
+ * This function prepares the initial and next operations to be generated using
+ * AsyncIterators, adding some extra metadata sunc as scope and prefix
+
+ */
+export function preparePageableOperations(
+  operationGroupDetails: OperationGroupDetails,
+  clientDetails: ClientDetails
+) {
+  if (clientDetails.options.disablePagingAsyncIterators) {
+    return;
+  }
+
+  operationGroupDetails.operations
+    .filter(o => o.pagination)
+    .forEach(operation => {
+      operation.scope = Scope.Private;
+      operation.namePrefix = "_";
+    });
+}
+
+/**
+ * This function generates all the required methods for the pageable operation
+ * using AsyncIterators. It generates 3 extra methods on top of the initial and next operations
+ * One public method, one private page method whcih gets results per page, and an All method that
+ * iterates throug pages, returning all results in a single collection
+ */
+export function writeAsyncIterators(
+  operationGroupDetails: OperationGroupDetails,
+  clientDetails: ClientDetails,
+  operationGroupClass: ClassDeclaration,
+  importedModels: Set<string>
+) {
+  if (clientDetails.options.disablePagingAsyncIterators) {
+    return;
+  }
+
+  operationGroupDetails.operations
+    // We can skip "next" operations since covering the original is enough. Otherwise we'll end up with duplicate methods.
+    .filter(o => o.pagination && !o.pagination?.isNextLinkMethod)
+    .forEach(operation => {
+      const initialOperationName = normalizeName(
+        operation.name,
+        NameType.Operation
+      );
+
+      const nextOperationName = normalizeName(
+        operation.pagination?.nextLinkOperationName || initialOperationName,
+        NameType.Operation
+      );
+
+      const nextOperation = operationGroupDetails.operations.find(
+        o =>
+          o.name.toLocaleLowerCase() ===
+          operation.pagination?.nextLinkOperationName?.toLocaleLowerCase()
+      );
+
+      // We need to figure out the body response type to be able to set the return types properly
+      const bodyResponseType = getPagingResponseBodyType(operation);
+
+      // Since we'll be using this type name for importing, we need to make sure that it is not in
+      // the form of array [], just need the name.
+      let bodyResponseTypeName =
+        bodyResponseType?.typeName.replace("[]", "") || "";
```

Comment:
I think the assumption here is that `bodyResponseType` is of the form `T[]` where `T` is a type name. I think we need to make sure this assumption is checked just in case. Otherwise, I would rather work with the type as a structure rather than as a string.

Reply:
This is used to import the required Model, removing [] is a safeguard to make sure we don't try to do `import {Model[]} from "./"`

If the typeName is not an array, it should be fine the replace will have no effect